File: examples/nfft/flags.c
/*
 * Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/*! \file flags.c
 *
 * \brief Testing the nfft.
 *
 * \author Stefan Kunis
 *
 * References: Time and Memory Requirements of the Nonequispaced FFT
 */
#include "config.h"

#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#ifdef HAVE_COMPLEX_H
#include <complex.h>
#endif

#include "nfft3.h"
#include "nfft3mp.h"
#include "nfft3util.h"

#ifdef GAUSSIAN
  unsigned test_fg=1;
#else
  unsigned test_fg=0;
#endif

#ifdef MEASURE_TIME_FFTW
  unsigned test_fftw=1;
#else
  unsigned test_fftw=0;
#endif

#ifdef MEASURE_TIME
  unsigned test=1;
#else
  unsigned test=0;
#endif

static void flags_cp(NFFT(plan) *dst, NFFT(plan) *src)
{
  dst->x = src->x;
  dst->f_hat = src->f_hat;
  dst->f = src->f;
  dst->g1 = src->g1;
  dst->g2 = src->g2;
  dst->my_fftw_plan1 = src->my_fftw_plan1;
  dst->my_fftw_plan2 = src->my_fftw_plan2;
}

static void time_accuracy(int d, int N, int M, int n, int m, unsigned test_ndft,
    unsigned test_pre_full_psi)
{
  int r, NN[d], nn[d];
  NFFT_R t_ndft, t, e;
  NFFT_C *swapndft = NULL;
  double t0, t1;

  NFFT(plan) p;
  NFFT(plan) p_pre_phi_hut;
  NFFT(plan) p_fg_psi;
  NFFT(plan) p_pre_lin_psi;
  NFFT(plan) p_pre_fg_psi;
  NFFT(plan) p_pre_psi;
  NFFT(plan) p_pre_full_psi;

  printf("%d\t%d\t", d, N);

  for (r = 0; r < d; r++)
  {
    NN[r] = N;
    nn[r] = n;
  }

  /* output vector ndft */
  if (test_ndft)
    swapndft = (NFFT_C*) NFFT(malloc)((size_t)(M) * sizeof(NFFT_C));

  NFFT(init_guru)(&p, d, NN, M, nn, m,
  MALLOC_X | MALLOC_F_HAT | MALLOC_F |
  FFTW_INIT | FFT_OUT_OF_PLACE,
  FFTW_MEASURE | FFTW_DESTROY_INPUT);

  /** init pseudo random nodes */
  NFFT(vrand_shifted_unit_double)(p.x, p.d * p.M_total);

  NFFT(init_guru)(&p_pre_phi_hut, d, NN, M, nn, m, PRE_PHI_HUT, 0);
  flags_cp(&p_pre_phi_hut, &p);
  NFFT(precompute_one_psi)(&p_pre_phi_hut);

  if (test_fg)
  {
    NFFT(init_guru)(&p_fg_psi, d, NN, M, nn, m, FG_PSI, 0);
    flags_cp(&p_fg_psi, &p);
    NFFT(precompute_one_psi)(&p_fg_psi);
  }

  NFFT(init_guru)(&p_pre_lin_psi, d, NN, M, nn, m, PRE_LIN_PSI, 0);
  flags_cp(&p_pre_lin_psi, &p);
  NFFT(precompute_one_psi)(&p_pre_lin_psi);

  if (test_fg)
  {
    NFFT(init_guru)(&p_pre_fg_psi, d, NN, M, nn, m, PRE_FG_PSI, 0);
    flags_cp(&p_pre_fg_psi, &p);
    NFFT(precompute_one_psi)(&p_pre_fg_psi);
  }

  NFFT(init_guru)(&p_pre_psi, d, NN, M, nn, m, PRE_PSI, 0);
  flags_cp(&p_pre_psi, &p);
  NFFT(precompute_one_psi)(&p_pre_psi);

  if (test_pre_full_psi)
  {
    NFFT(init_guru)(&p_pre_full_psi, d, NN, M, nn, m, PRE_FULL_PSI, 0);
    flags_cp(&p_pre_full_psi, &p);
    NFFT(precompute_one_psi)(&p_pre_full_psi);
  }

  /* init pseudo random Fourier coefficients */
  NFFT(vrand_unit_complex)(p.f_hat, p.N_total);

  /* NDFT */
  if (test_ndft)
  {
    NFFT_CSWAP(p.f, swapndft);

    t_ndft = NFFT_K(0.0);
    r = 0;
    while (t_ndft < NFFT_K(0.01))
    {
      r++;
      t0 = NFFT(clock_gettime_seconds)();
      NFFT(trafo_direct)(&p);
      t1 = NFFT(clock_gettime_seconds)();
      t = t1 - t0;
      t_ndft += t;
    }
    t_ndft /= (NFFT_R)(r);

    NFFT_CSWAP(p.f, swapndft);
  }
  else
    t_ndft = NFFT_MKNAN("");

  /* NFFTs */
  NFFT(trafo)(&p);
  NFFT(trafo)(&p_pre_phi_hut);
  if (test_fg)
    NFFT(trafo)(&p_fg_psi);
  else
    p_fg_psi.MEASURE_TIME_t[2] = NFFT_MKNAN("");
  NFFT(trafo)(&p_pre_lin_psi);
  if (test_fg)
    NFFT(trafo)(&p_pre_fg_psi);
  else
    p_pre_fg_psi.MEASURE_TIME_t[2] = NFFT_MKNAN("");
  NFFT(trafo)(&p_pre_psi);
  if (test_pre_full_psi)
    NFFT(trafo)(&p_pre_full_psi);
  else
    p_pre_full_psi.MEASURE_TIME_t[2] = NFFT_MKNAN("");

  if (test_fftw == 0)
    p.MEASURE_TIME_t[1] = NFFT_MKNAN("");

  if (test_ndft)
    e = NFFT(error_l_2_complex)(swapndft, p.f, p.M_total);
  else
    e = NFFT_MKNAN("");

  printf(
      "%.2" NFFT__FES__ "\t%d\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\t%.2" NFFT__FES__ "\n",
      t_ndft, m, e, p.MEASURE_TIME_t[0], p_pre_phi_hut.MEASURE_TIME_t[0],
      p.MEASURE_TIME_t[1], p.MEASURE_TIME_t[2], p_fg_psi.MEASURE_TIME_t[2],
      p_pre_lin_psi.MEASURE_TIME_t[2], p_pre_fg_psi.MEASURE_TIME_t[2],
      p_pre_psi.MEASURE_TIME_t[2], p_pre_full_psi.MEASURE_TIME_t[2]);

  fflush(stdout);

  /** finalise */
  if (test_pre_full_psi)
    NFFT(finalize)(&p_pre_full_psi);
  NFFT(finalize)(&p_pre_psi);
  if (test_fg)
    NFFT(finalize)(&p_pre_fg_psi);
  NFFT(finalize)(&p_pre_lin_psi);
  if (test_fg)
    NFFT(finalize)(&p_fg_psi);
  NFFT(finalize)(&p_pre_phi_hut);
  NFFT(finalize)(&p);

  if (test_ndft)
    NFFT(free)(swapndft);
}

static void accuracy_pre_lin_psi(int d, int N, int M, int n, int m, int K)
{
  int r, NN[d], nn[d];
  NFFT_R e;
  NFFT_C *swapndft;

  NFFT(plan) p;

  for (r = 0; r < d; r++)
  {
    NN[r] = N;
    nn[r] = n;
  }

  /* output vector ndft */
  swapndft = (NFFT_C*) NFFT(malloc)((size_t)(M) * sizeof(NFFT_C));

  NFFT(init_guru)(&p, d, NN, M, nn, m,
  MALLOC_X | MALLOC_F_HAT | MALLOC_F |
  PRE_PHI_HUT | PRE_LIN_PSI |
  FFTW_INIT | FFT_OUT_OF_PLACE,
  FFTW_MEASURE | FFTW_DESTROY_INPUT);

  /** realloc psi */
  NFFT(free)(p.psi);
  p.K = K;
  p.psi = (NFFT_R*) NFFT(malloc)((size_t)((p.K + 1) * p.d) * sizeof(NFFT_R));

  /** precomputation can be done before the nodes are initialised */
  NFFT(precompute_one_psi)(&p);

  /** init pseudo random nodes */
  NFFT(vrand_shifted_unit_double)(p.x, p.d * p.M_total);

  /** init pseudo random Fourier coefficients */
  NFFT(vrand_unit_complex)(p.f_hat, p.N_total);

  /** compute exact result */
  NFFT_CSWAP(p.f, swapndft);
  NFFT(trafo_direct)(&p);
  NFFT_CSWAP(p.f, swapndft);

  /** NFFT */
  NFFT(trafo)(&p);
  e = NFFT(error_l_2_complex)(swapndft, p.f, p.M_total);

  //  printf("%d\t%d\t%d\t%d\t%.2e\n",d,N,m,K,e);
  printf("$%.1" NFFT__FES__ "$&\t", e);

  fflush(stdout);

  /** finalise */
  NFFT(finalize)(&p);
  NFFT(free)(swapndft);
}

int main(int argc, char **argv)
{
  int l, trial;

  if (argc <= 2)
  {
    fprintf(stderr, "flags type first last trials d m\n");
    return EXIT_FAILURE;
  }

  if ((test == 0) && (atoi(argv[1]) < 2))
  {
    fprintf(stderr, "configure with --enable-measure-time\n");
    return EXIT_FAILURE;
  }

  fprintf(stderr, "Testing different precomputation schemes for the nfft.\n");
  fprintf(stderr, "Columns: d, N=M, t_ndft, e_nfft, t_D, t_pre_phi_hut, ");
  fprintf(stderr, "t_fftw, t_B, t_fg_psi, t_pre_lin_psi, t_pre_fg_psi, ");
  fprintf(stderr, "t_pre_psi, t_pre_full_psi\n\n");

  int arg2 = atoi(argv[2]);
  int arg3 = atoi(argv[3]);
  int arg4 = atoi(argv[4]);

  /* time vs. N=M */
  if (atoi(argv[1]) == 0)
  {
    int d = atoi(argv[5]);
    int m = atoi(argv[6]);

    for (l = arg2; l <= arg3; l++)
    {
      int N = (int)(1U << l);
      int M = (int)(1U << (d * l));
      for (trial = 0; trial < arg4; trial++)
      {
        time_accuracy(d, N, M, 2 * N, m, 0, 0);
      }
    }
  }
  else if (atoi(argv[1]) == 1) /* accuracy vs. time */
  {
    int d = atoi(argv[5]);
    int N = atoi(argv[6]);
    int m;

    for (m = arg2; m <= arg3; m++)
    {
      for (trial = 0; trial < arg4; trial++)
      {
        time_accuracy(d, N, (int)(NFFT_LRINT(NFFT_POW((NFFT_R)(N), (NFFT_R)(d)))), 2 * N, m, 1, 1);
      }
    }
  }
  else if (atoi(argv[1]) == 2) /* accuracy vs. K for linear interpolation, assumes (m+1)|K */
  {
    int d = atoi(argv[5]);
    int N = atoi(argv[6]);
    int m = atoi(argv[7]);

    printf("$\\log_2(K/(m+1))$&\t");

    for (l = arg2; l < arg3; l++)
      printf("$%d$&\t", l);

    printf("$%d$\\\\\n", arg3);

    printf("$\\tilde E_2$&\t");
    for (l = arg2; l <= arg3; l++)
    {
      int x = (m + 1) * (int)(1U << l);
      accuracy_pre_lin_psi(d, N, (int)(NFFT_LRINT(NFFT_POW((NFFT_R)(N), (NFFT_R)(d)))), 2 * N, m, x);
    }

    printf("\n");
  }


  return EXIT_SUCCESS;
}

File: examples/nfft/flags.m
% Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
%
% This program is free software; you can redistribute it and/or modify it under
% the terms of the GNU General Public License as published by the Free Software
% Foundation; either version 2 of the License, or (at your option) any later
% version.
%
% This program is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
% FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
% details.
%
% You should have received a copy of the GNU General Public License along with
% this program; if not, write to the Free Software Foundation, Inc., 51
% Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
%
%% File: flags1.m
%%
%% Testing different precomputation schemes for the nfft.
%% 1. Computation time vs. problem size
%%
%% Author Stefan Kunis
%%
%% References: Time and memory requirements of the Nonequispaced FFT
%%
%% Loads flags.data0.gaussian, ...

% see flags.readme
trials=10;
first=4;
last=20;

% load already computed data, times for precomputed D/B differ only negligible
dg=load('flags.data0.gaussian');
dk=load('flags.data0.kaiser_bessel');
ds=load('flags.data0.sinc_power');
db=load('flags.data0.b_spline');

N=dg(1:trials:end,2);

t_Dg=(mean(reshape(dg(:,6),trials,last-first+1)))';
t_Dk=(mean(reshape(dk(:,6),trials,last-first+1)))';
t_Ds=(mean(reshape(ds(:,6),trials,last-first+1)))';
t_Db=(mean(reshape(db(:,6),trials,last-first+1)))';
t_pre_phi_hut=(mean(reshape(dg(:,7),trials,last-first+1)))';

t_fftw=(mean(reshape(dg(:,8),trials,last-first+1)))';

t_Bg=(mean(reshape(dg(:,9),trials,last-first+1)))';
t_Bk=(mean(reshape(dk(:,9),trials,last-first+1)))';
t_Bs=(mean(reshape(ds(:,9),trials,last-first+1)))';
t_Bb=(mean(reshape(db(:,9),trials,last-first+1)))';

t_fg_psi=(mean(reshape(dg(:,10),trials,last-first+1)))';
t_pre_lin_psi=(mean(reshape(dg(:,11),trials,last-first+1)))';
t_pre_fg_psi=(mean(reshape(dg(:,12),trials,last-first+1)))';
t_pre_psi=(mean(reshape(dg(:,13),trials,last-first+1)))';
t_pre_full_psi=(mean(reshape(dg(:,14),trials,last-first+1)))';

h=loglog(N,t_Dg,'k',...
         N,t_Dk,'k--',...
         N,t_Ds,'k-.',...
         N,t_Db,'k:',...
         N,t_pre_phi_hut,'k.');
  
set(h,'LineWidth',1.8); set(h,'MarkerSize',6); 
set(gca,'YTick',[10^-6,10^-4,10^-2,1,10^2]);
set(gca,'XTick',[10^2,10^3,10^4,10^5,10^6]);
set(gca,'FontSize',20);
axis([N(1),N(end),10^-6,10^1]);

if(to_pdf)
  !ps2pdf temp.eps flags1.pdf 
  !rm temp.eps
else
  !mv temp.eps flags1.eps
end;

return

%%
%% Testing accuracy vs. cut-off/Taylor degree m.
%%
trials=20;
first=1;
last=20;
sigma_nfft=2;
sigma_taylor=4;
sigma_nfft1=1.25;
sigma_taylor1=1.5;
sigma_nfft2=8;
sigma_taylor2=16;

% typical sigma
if ispc
    cmd='taylor_nfft.exe';
else 
    cmd='./taylor_nfft';
end
system(sprintf('%s %d %d %d %d %f %f > taylor_nfft2a.dat',cmd,1,first,...
	       last,trials,sigma_nfft,sigma_taylor));
data=load('taylor_nfft2a.dat');

m_nfft=data(:,5);
e_nfft=data(:,7);
max_e_nfft=(max(reshape(e_nfft,trials,last-first+1)))';

m_taylor=data(:,9);
e_taylor=data(:,11);
max_e_taylor=(max(reshape(e_taylor,trials,last-first+1)))';

% small sigma
system(sprintf('%s %d %d %d %d %f %f > taylor_nfft2b.dat',cmd,1,first,...
	       last,trials,sigma_nfft1,sigma_taylor1));
data1=load('taylor_nfft2b.dat');
max_e_nfft_sigma=(max(reshape(data1(:,7),trials,last-first+1)))';
max_e_taylor_sigma=(max(reshape(data1(:,11),trials,last-first+1)))';

% large sigma
system(sprintf('%s %d %d %d %d %f %f > taylor_nfft2c.dat',cmd,1,first,...
	       last,trials,sigma_nfft2,sigma_taylor2));
data2=load('taylor_nfft2c.dat');
max_e_nfft_sigma2=(max(reshape(data2(:,7),trials,last-first+1)))';
max_e_taylor_sigma2=(max(reshape(data2(:,11),trials,last-first+1)))';

h=semilogy(m_nfft,e_nfft,'k+',first:last,max_e_nfft,'k--',first:last,...
	   max_e_nfft_sigma,'k:',first:last,max_e_nfft_sigma2,'k:',...
	   m_taylor,e_taylor,'kx',first:last,max_e_taylor,'k',first:last,...
	   max_e_taylor_sigma,'k-.',first:last,max_e_taylor_sigma2,'k-.');
set(h,'LineWidth',1.8); set(h,'MarkerSize',6); 
set(gca,'YTick',[10^-15,10^-10,10^-5,1]);
set(gca,'FontSize',20);
axis([first,last,10^-16,1]);

print temp.eps -deps
!ps2pdf temp.eps taylor_nfft2.pdf 
!rm temp.eps


%%
%% Testing accuracy vs. time.
%%
t_nfft=data(:,6);
t_taylor=data(:,10);

h=semilogy(t_nfft,e_nfft,'k+',t_taylor,e_taylor,'kx');
set(h,'LineWidth',1.8); set(h,'MarkerSize',6); 
set(gca,'YTick',[10^-15,10^-10,10^-5,1]);
set(gca,'FontSize',20);
axis([min([t_nfft;t_taylor]),max([t_nfft;t_taylor]),10^-16,1]);

print temp.eps -deps
!ps2pdf temp.eps taylor_nfft3.pdf 
!rm temp.eps

